azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSv2FileIO.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.azure.adlsv2;
+
+import com.azure.core.http.HttpClient;
+import com.azure.storage.file.datalake.DataLakeFileClient;
+import com.azure.storage.file.datalake.DataLakePathClientBuilder;
+import java.util.Map;
+import org.apache.iceberg.azure.AzureProperties;
+import org.apache.iceberg.common.DynConstructors;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.metrics.MetricsContext;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.util.SerializableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * FileIO implementation backed by Azure Data Lake Storage v2 (ADLSv2)
+ *
+ * <p>Locations follow the conventions used by Hadoop's Azure support, i.e.
+ *
+ * <pre>{@code abfs[s]://<container>@<storage account>.dfs.core.windows.net/<blob_path>}</pre>
+ *
+ * <p>See <a href="https://hadoop.apache.org/docs/stable/hadoop-azure/abfs.html">Hadoop Azure
+ * Support</a>
+ */
+public class ADLSv2FileIO implements FileIO {
+  private static final Logger LOG = LoggerFactory.getLogger(ADLSv2FileIO.class);
+  private static final String DEFAULT_METRICS_IMPL =
+      "org.apache.iceberg.hadoop.HadoopMetricsContext";
+
+  private static final HttpClient HTTP = HttpClient.createDefault();
+
+  private AzureProperties azureProperties;
+  private MetricsContext metrics = MetricsContext.nullMetrics();
+  private SerializableMap<String, String> properties = null;
+
+  /**
+   * No-arg constructor to load the FileIO dynamically.
+   *
+   * <p>All fields are initialized by calling {@link ADLSv2FileIO#initialize(Map)} later.
+   */
+  public ADLSv2FileIO() {}
+
+  /**
+   * Constructor with Azure properties.
+   *
+   * <p>Calling {@link ADLSv2FileIO#initialize(Map)} will overwrite information set in this
+   * constructor.
+   *
+   * @param azureProperties Azure properties
+   */
+  public ADLSv2FileIO(AzureProperties azureProperties) {
```

Comment:
do we need this one? It doesn't seem to be used

Reply:
It isn't really needed anymore as we're using initialize() in the tests. I removed it.